DOCS-10592-jms-ack-operation-kt (#1402)
mulesoft/docs-connectors: jms/1.7/modules/ROOT/pages/jms-ack.adoc

= Configure Message Acknowledgement with the JMS Connector
:keywords: jms, connector, consume, message, ack
:page-aliases: connectors::jms/jms-ack.adoc

> Anypoint Connector for JMS (JMS Connector) provides different message acknowledgement modes to configure for the *Ack*, *Consume*, *On New Message* and *Recovery session* operations:

Reviewer: series comma:
```suggestion
Anypoint Connector for JMS (JMS Connector) provides different message acknowledgement modes to configure for the *Ack*, *Consume*, *On New Message*, and *Recovery session* operations:
```

* Automatic acknowledgement on success
* Immediate acknowledgement
* Manual acknowledgement

> The following examples illustrate how to configure the modes for the operations.

Reviewer: ```suggestion
The following examples illustrate how to configure each message acknowledgement mode.
```

> == Configure Automatic Acknowledgement On Success Mode

Reviewer: ```suggestion
== Configure Automatic Acknowledgement on Success Mode
```

Configure the automatic acknowledgement on success mode in the *On New Message* source by setting the *Acknowledge Mode* field to either one of these modes:

> * *AUTO* +

Reviewer: Why is this mode name in boldface? Is it a UI label?

> Automatically acknowledges (AKC) the received message only if the flow execution completes successfully. In case of an error, the message is not acknowledged and the redelivery occurs.

Reviewer: (AKC) > (ACK)

Reviewer: Just an alternative, not a correction:
```suggestion
Automatically acknowledges (ACK) the received message only if the flow execution completes successfully; otherwise, the message is not acknowledged and is redelivered.
```

> * *DUPS_OK* +

Reviewer: Why bold? GLOBAL for mode names.

> Automatically acknowledges (AKC) the received message only if the flow execution completes successfully but may lead to duplicates because the message may be redelivered before the ACK is performed.

Reviewer: (ACK)

Reviewer: ```suggestion
Like *AUTO* mode, automatically acknowledges (ACK) the received message only if the flow execution completes successfully, but leads to duplicate messages if the message is redelivered before the ACK is performed.
```

In the following example, you configure the *AUTO* acknowledgement mode:

. In Studio, select the *On New Message* source from your flow.
. In the *On New Message* configuration screen, in *Destination*, specify the name of the destination from which to consume the message, for example, `#[vars.destination]`.
. Set the *Acknowledge Mode* field to *AUTO*.

.JMS Connector Auto Acknowledgement Mode configuration
image::jms-ack-auto.png[In the On New Message source screen, set the Acknowledge Mode field to AUTO]

> In the XML editor, the `<jms:listener>` and `ackMode` configurations looks like this:

Reviewer: ```suggestion
In the XML editor, the `<jms:listener>` and `ackMode` configurations look like this:
```

[source,xml,linenums]
----
<jms:listener  config-ref="JMS_Config" destination="#[vars.destination]" ackMode="AUTO"/>
----

== Configure Immediate Acknowledgement Mode

> Configure the immediate acknowledgement mode in the *On New Message* source by using the *Acknowledge Mode* field, or in the *Consume* operation by using the *Ack mode* field. Set these fields to *IMMEDIATE*. This mode automatically acknowledges (AKC) the message once it is consumed and prior to any application processing of the message. +

Reviewer: (ACK)

> Configure the immediate acknowledgement mode in the *On New Message* source by using the *Acknowledge Mode* field, or in the *Consume* operation by using the *Ack mode* field. Set these fields to *IMMEDIATE*. This mode automatically acknowledges (AKC) the message once it is consumed and prior to any application processing of the message. +

Reviewer: Change "these" to "the" because you are using only one of the two fields, right?
```suggestion
Configure the immediate acknowledgement mode in the *On New Message* source by using the *Acknowledge Mode* field, or in the *Consume* operation by using the *Ack mode* field. Set the fields to *IMMEDIATE*. This mode automatically acknowledges (ACK) the message after it is consumed and before any application processing of the message. +
```
Changed "prior" only because the antonym to "after" is "before."

Reviewer: "Set the field": Is this a field that has a menu from which you select or do you type in the field? Related to why ACK modes are bold....

> Automatically acknowledging the message once it is received indicates that the message won't be redelivered if any error occurs during the processing of the message, so an application logic like a dead letter queue has to be created to manage errors without losing messages.

Reviewer: once > after, if that is what you really mean (this is a translation issue, because "once" could be misread as "one time.")
```suggestion
The result of automatically acknowledging the message is that the message won't be redelivered if any error occurs during message processing, requiring application logic like a "dead letter" queue to manage errors without losing messages.
```

In the following example, you configure the *IMMEDIATE* acknowledgement mode for the *Consume* operation:

. In Studio, select the *Consume* operation from your flow.
. In the *Consume* configuration screen, in *Destination*, specify the name of the destination from which to consume the message, for example, `#[vars.destination]`.

> . Set the *Ack mode* field to *IMMEDIATE*.

Reviewer: ditto comment Line 41 and throughout

.JMS Connector Immediate Acknowledgement Mode configuration
image::jms-ack-immediate.png[In the Consume operation screen, set the Ack Mode field to IMMEDIATE]

> In the XML editor, the `<jms:consume>` and `ackMode` configurations looks like this:

Reviewer: ```suggestion
In the XML editor, the `<jms:consume>` and `ackMode` configurations look like this:
```

== Configure Manual Acknowledgement Mode

> Configure the manual acknowledgement mode in the *On New Message* source by using the *Acknowledge Mode* field, or in the *Consume* operation by using the *Ack mode* field. Set these fields to *MANUAL*. This mode delegates all the responsibility of performing the acknowledgement (ACK) on the message to the application logic.

Reviewer: ```suggestion
Configure the manual acknowledgement mode in the *On New Message* source by using the *Acknowledge Mode* field, or in the *Consume* operation by using the *Ack mode* field. Set these fields to *MANUAL*. This mode delegates all responsibility for performing message acknowledgement (ACK) to the application logic.
```

> Every message received by either the *On New Message* or *Consume* operations have an acknowledgement ID available in the Mule message attributes which identifies the message uniquely for a given connection. +

Reviewer: ```suggestion
Every message received by either the *On New Message* or *Consume* operations includes an acknowledgement ID, the Mule message attributes, that identifies the message uniquely for a given connection. +
```

The acknowledgement ID that identifies the message is then passed to the *Ack* operation in the *Ack id* field.

In the following example, you configure the *MANUAL* acknowledgement mode for the *Consume* operation:

. In Studio, select the *Consume* operation from your flow.
. In the *Consume* configuration screen, in *Destination*, specify the name of the destination from which to consume the message, for example, `openTickets`.
. Set the *Ack mode* field to *MANUAL*.
. Drag a *Logger* component to the right of the *Consume* operation.
. Set the *Message* field to `#[payload]`.
. Drag an *Ack* operation to the right of the *Logger* component.
. Set the *Ack id* field to the acknowledgement ID of the message session to recover, for example, `#[attributes.ackId]`.

.JMS Connector Ack operation configuration
image::jms-ack-operation.png[In the Ack operation screen, set the Ack id field to the acknowledgement ID of the message session to recover]

> In the XML editor, the `<jms:consume>`, `ackMode`, and `ackId` configurations looks like this:

Reviewer: ```suggestion
In the XML editor, the `<jms:consume>`, `ackMode`, and `ackId` configurations look like this:
```

----

=== Configure Manual Session Recovery

> When using the manual acknowledgement mode, all received messages and not acknowledged are not redelivered by the broker.

Reviewer: "all received messages" is not the actor "using" the mode. We need to reword to be grammatically correct, something like...:
```suggestion
With manual acknowledgement mode, all messages received but not acknowledged are not redelivered by the broker.
```

> In case of an error during the message processing, you can manually recover all the messages in the session that have to be redelivered using the *Recover session* operation. The connection used to recover the session has to be the the same as the connection for receiving the message.

Reviewer: ```suggestion
If there is an error during the message processing, you must manually recover all the messages in the session that have to be redelivered by using the *Recover session* operation. The connection used to recover the session has to be the the same as the connection for receiving the message.
```

. In Studio, select the *Consume* operation from your flow.
. In the *Consume* configuration screen, in *Destination*, specify the name of the destination from which to consume the message, for example, `${destination}`.
. Set the *Ack mode* field to *MANUAL*.
. Drag a *Logger* component to the right of the *Consume* operation.
. Set the *Message* field to `#[payload]`.
. Drag an *Ack* operation to the right of the *Logger* component.
. Set the *Ack id* field to the acknowledgement ID of the message session to recover, for example, `#[vars.consumedMessage.attributes.ackId]`.
. In your flow, expand the *Error handling* section by clicking the small arrow.
. Drag an *On Error Propagate* component in the *Error handling* section.
. Drag a *Recover session* operation inside the *On Error Propagate* component.
. Set the *Ack id* field to `#[vars.consumedMessage.attributes.ackId]`.

.JMS Connector Recover session operation configuration
image::jms-ack-recover.png[In the Recover session operation screen, set the Ack id field to the acknowledgement ID of the message session to recover]

> In the XML editor, the `<jms:consume>`, `ackMode`, and `ackId` configurations looks like this:

Reviewer: ```suggestion
In the XML editor, the `<jms:consume>`, `ackMode`, and `ackId` configurations look like this:
```

</flow>
----

== See Also

* xref:jms-topic-subscription.adoc[Configure Topic Subscriptions]

> * xref:jms-listener.adoc[Listen For New Messages]

Reviewer: For > for
